File: mobile/mobile_central/test/services/modules/drivers/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/drivers/app/use_cases.dart';
import 'package:mobile_central/services/modules/drivers/domain/entities.dart';
import 'package:mobile_central/services/modules/drivers/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// --- Manual Mock ---

class MockDriverRepository implements IDriverRepository {
  final List<String> calls = [];

  PaginatedResponse<DriverInfo>? getDriversResult;
  DriverInfo? getDriverByIdResult;
  DriverInfo? createDriverResult;
  DriverInfo? updateDriverResult;
  Map<String, dynamic>? deleteDriverResult;

  Exception? errorToThrow;

  GetDriversParams? capturedGetDriversParams;
  int? capturedId;
  int? capturedBusinessId;
  CreateDriverDTO? capturedCreateData;
  int? capturedUpdateId;
  UpdateDriverDTO? capturedUpdateData;
  int? capturedDeleteId;

  @override
  Future<PaginatedResponse<DriverInfo>> getDrivers(GetDriversParams? params) async {
    calls.add('getDrivers');
    capturedGetDriversParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getDriversResult!;
  }

  @override
  Future<DriverInfo> getDriverById(int id, {int? businessId}) async {
    calls.add('getDriverById');
    capturedId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getDriverByIdResult!;
  }

  @override
  Future<DriverInfo> createDriver(CreateDriverDTO data, {int? businessId}) async {
    calls.add('createDriver');
    capturedCreateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createDriverResult!;
  }

  @override
  Future<DriverInfo> updateDriver(int id, UpdateDriverDTO data, {int? businessId}) async {
    calls.add('updateDriver');
    capturedUpdateId = id;
    capturedUpdateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return updateDriverResult!;
  }

  @override
  Future<Map<String, dynamic>> deleteDriver(int id, {int? businessId}) async {
    calls.add('deleteDriver');
    capturedDeleteId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return deleteDriverResult ?? {'message': 'deleted'};
  }
}

// --- Helpers ---

DriverInfo _makeDriver({int id = 1, String firstName = 'TestDriver'}) {
  return DriverInfo(
    id: id,
    businessId: 1,
    firstName: firstName,
    lastName: 'Last',
    email: 'test@test.com',
    phone: '555-1234',
    identification: 'CC-123',
    status: 'active',
    photoUrl: '',
    licenseType: 'B1',
    createdAt: '2026-01-01',
    updatedAt: '2026-01-02',
  );
}

Pagination _makePagination() {
  return Pagination(
    currentPage: 1,
    perPage: 10,
    total: 1,
    lastPage: 1,
    hasNext: false,
    hasPrev: false,
  );
}

// --- Tests ---

void main() {
  late MockDriverRepository mockRepo;
  late DriverUseCases useCases;

  setUp(() {
    mockRepo = MockDriverRepository();
    useCases = DriverUseCases(mockRepo);
  });

  group('getDrivers', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<DriverInfo>(
        data: [_makeDriver()],
        pagination: _makePagination(),
      );
      mockRepo.getDriversResult = expected;
      final params = GetDriversParams(page: 1, pageSize: 10);

      final result = await useCases.getDrivers(params);

      expect(result.data.length, 1);
      expect(result.data[0].firstName, 'TestDriver');
      expect(mockRepo.calls, ['getDrivers']);
      expect(mockRepo.capturedGetDriversParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getDriversResult = PaginatedResponse<DriverInfo>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getDrivers(null);

      expect(mockRepo.capturedGetDriversParams, isNull);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Network error');

      expect(() => useCases.getDrivers(null), throwsException);
    });
  });

  group('getDriverById', () {
    test('delegates to repository with correct id', () async {
      mockRepo.getDriverByIdResult = _makeDriver(id: 42, firstName: 'Found');

      final result = await useCases.getDriverById(42);

      expect(result.id, 42);
      expect(result.firstName, 'Found');
      expect(mockRepo.capturedId, 42);
      expect(mockRepo.calls, ['getDriverById']);
    });

    test('passes businessId to repository', () async {
      mockRepo.getDriverByIdResult = _makeDriver();

      await useCases.getDriverById(1, businessId: 5);

      expect(mockRepo.capturedBusinessId, 5);
    });
  });

  group('createDriver', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateDriverDTO(
        firstName: 'New',
        lastName: 'Driver',
        phone: '555',
        identification: 'CC-999',
      );
      mockRepo.createDriverResult = _makeDriver(id: 99, firstName: 'New');

      final result = await useCases.createDriver(dto);

      expect(result.id, 99);
      expect(result.firstName, 'New');
      expect(mockRepo.capturedCreateData, dto);
      expect(mockRepo.calls, ['createDriver']);
    });

    test('passes businessId to repository', () async {
      final dto = CreateDriverDTO(
        firstName: 'Test',
        lastName: 'Test',
        phone: '555',
        identification: 'CC-1',
      );
      mockRepo.createDriverResult = _makeDriver();

      await useCases.createDriver(dto, businessId: 7);

      expect(mockRepo.capturedBusinessId, 7);
    });
  });

  group('updateDriver', () {
    test('delegates to repository with correct id and data', () async {
      final dto = UpdateDriverDTO(firstName: 'Updated');
      mockRepo.updateDriverResult = _makeDriver(id: 5, firstName: 'Updated');

      final result = await useCases.updateDriver(5, dto);

      expect(result.id, 5);
      expect(result.firstName, 'Updated');
      expect(mockRepo.capturedUpdateId, 5);
      expect(mockRepo.capturedUpdateData, dto);
      expect(mockRepo.calls, ['updateDriver']);
    });

    test('passes businessId to repository', () async {
      final dto = UpdateDriverDTO(firstName: 'Test');
      mockRepo.updateDriverResult = _makeDriver();

      await useCases.updateDriver(1, dto, businessId: 3);

      expect(mockRepo.capturedBusinessId, 3);
    });
  });

  group('deleteDriver', () {
    test('delegates to repository with correct id', () async {
      mockRepo.deleteDriverResult = {'message': 'deleted'};

      final result = await useCases.deleteDriver(7);

      expect(result['message'], 'deleted');
      expect(mockRepo.capturedDeleteId, 7);
      expect(mockRepo.calls, ['deleteDriver']);
    });

    test('passes businessId to repository', () async {
      await useCases.deleteDriver(7, businessId: 2);

      expect(mockRepo.capturedBusinessId, 2);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Not found');

      expect(() => useCases.deleteDriver(7), throwsException);
    });
  });
}

File: front/central/src/services/modules/drivers/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { DriverUseCases } from './use-cases';
import { IDriverRepository } from '../domain/ports';
import {
    DriverInfo,
    DriversListResponse,
    DeleteDriverResponse,
    CreateDriverDTO,
    UpdateDriverDTO,
} from '../domain/types';

// -----------------------------------------------------------------
// Helpers: datos de prueba reutilizables
// -----------------------------------------------------------------

const makeDriver = (overrides: Partial<DriverInfo> = {}): DriverInfo => ({
    id: 1,
    business_id: 1,
    first_name: 'Juan',
    last_name: 'Perez',
    email: 'juan.perez@example.com',
    phone: '+573001234567',
    identification: '1234567890',
    status: 'active',
    photo_url: '',
    license_type: 'B1',
    license_expiry: '2027-12-31',
    warehouse_id: 1,
    notes: null,
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeDriversListResponse = (overrides: Partial<DriversListResponse> = {}): DriversListResponse => ({
    data: [makeDriver()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
    ...overrides,
});

const deleteSuccess: DeleteDriverResponse = { message: 'Conductor eliminado exitosamente' };
const deleteError: DeleteDriverResponse = { error: 'Conductor no encontrado' };

// -----------------------------------------------------------------
// Mock del repositorio
// -----------------------------------------------------------------

function createMockRepository(): IDriverRepository {
    return {
        getDrivers: vi.fn(),
        getDriverById: vi.fn(),
        createDriver: vi.fn(),
        updateDriver: vi.fn(),
        deleteDriver: vi.fn(),
    } as unknown as IDriverRepository;
}

// -----------------------------------------------------------------
// Suite principal
// -----------------------------------------------------------------

describe('DriverUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: DriverUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new DriverUseCases(repo as unknown as IDriverRepository);
    });

    // ---------------------------------------------------------------
    // getDrivers
    // ---------------------------------------------------------------
    describe('getDrivers', () => {
        it('debería retornar la lista paginada de conductores cuando el repositorio tiene éxito', async () => {
            const response = makeDriversListResponse();
            vi.mocked(repo.getDrivers).mockResolvedValue(response);

            const result = await useCases.getDrivers({ page: 1, page_size: 10 });

            expect(result).toEqual(response);
            expect(repo.getDrivers).toHaveBeenCalledOnce();
            expect(repo.getDrivers).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            const response = makeDriversListResponse();
            vi.mocked(repo.getDrivers).mockResolvedValue(response);

            await useCases.getDrivers();

            expect(repo.getDrivers).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Fallo de base de datos');
            vi.mocked(repo.getDrivers).mockRejectedValue(expectedError);

            await expect(useCases.getDrivers()).rejects.toThrow('Fallo de base de datos');
        });
    });

    // ---------------------------------------------------------------
    // getDriverById
    // ---------------------------------------------------------------
    describe('getDriverById', () => {
        it('debería retornar un conductor por su ID', async () => {
            const driver = makeDriver();
            vi.mocked(repo.getDriverById).mockResolvedValue(driver);

            const result = await useCases.getDriverById(1);

            expect(result).toEqual(driver);
            expect(repo.getDriverById).toHaveBeenCalledOnce();
            expect(repo.getDriverById).toHaveBeenCalledWith(1, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const driver = makeDriver();
            vi.mocked(repo.getDriverById).mockResolvedValue(driver);

            await useCases.getDriverById(1, 5);

            expect(repo.getDriverById).toHaveBeenCalledWith(1, 5);
        });

        it('debería propagar el error cuando el conductor no existe', async () => {
            const expectedError = new Error('Conductor no encontrado');
            vi.mocked(repo.getDriverById).mockRejectedValue(expectedError);

            await expect(useCases.getDriverById(999)).rejects.toThrow('Conductor no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // createDriver
    // ---------------------------------------------------------------
    describe('createDriver', () => {
        const dto: CreateDriverDTO = {
            first_name: 'Carlos',
            last_name: 'Lopez',
            phone: '+573009876543',
            identification: '9876543210',
            license_type: 'C1',
        };

        it('debería crear un conductor y retornar la respuesta del repositorio', async () => {
            const createdDriver = makeDriver({ first_name: 'Carlos', last_name: 'Lopez' });
            vi.mocked(repo.createDriver).mockResolvedValue(createdDriver);

            const result = await useCases.createDriver(dto);

            expect(result).toEqual(createdDriver);
            expect(repo.createDriver).toHaveBeenCalledOnce();
            expect(repo.createDriver).toHaveBeenCalledWith(dto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const createdDriver = makeDriver();
            vi.mocked(repo.createDriver).mockResolvedValue(createdDriver);

            await useCases.createDriver(dto, 3);

            expect(repo.createDriver).toHaveBeenCalledWith(dto, 3);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            const expectedError = new Error('Identificación duplicada');
            vi.mocked(repo.createDriver).mockRejectedValue(expectedError);

            await expect(useCases.createDriver(dto)).rejects.toThrow('Identificación duplicada');
        });
    });

    // ---------------------------------------------------------------
    // updateDriver
    // ---------------------------------------------------------------
    describe('updateDriver', () => {
        const updateDto: UpdateDriverDTO = { first_name: 'Carlos Actualizado', phone: '+573001111111' };

        it('debería actualizar un conductor y retornar la respuesta del repositorio', async () => {
            const updatedDriver = makeDriver({ first_name: 'Carlos Actualizado' });
            vi.mocked(repo.updateDriver).mockResolvedValue(updatedDriver);

            const result = await useCases.updateDriver(1, updateDto);

            expect(result).toEqual(updatedDriver);
            expect(repo.updateDriver).toHaveBeenCalledOnce();
            expect(repo.updateDriver).toHaveBeenCalledWith(1, updateDto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const updatedDriver = makeDriver();
            vi.mocked(repo.updateDriver).mockResolvedValue(updatedDriver);

            await useCases.updateDriver(1, updateDto, 5);

            expect(repo.updateDriver).toHaveBeenCalledWith(1, updateDto, 5);
        });

        it('debería propagar el error cuando la actualización falla', async () => {
            const expectedError = new Error('Conductor no encontrado');
            vi.mocked(repo.updateDriver).mockRejectedValue(expectedError);

            await expect(useCases.updateDriver(99, updateDto)).rejects.toThrow('Conductor no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // deleteDriver
    // ---------------------------------------------------------------
    describe('deleteDriver', () => {
        it('debería eliminar un conductor y retornar confirmación', async () => {
            vi.mocked(repo.deleteDriver).mockResolvedValue(deleteSuccess);

            const result = await useCases.deleteDriver(1);

            expect(result).toEqual(deleteSuccess);
            expect(repo.deleteDriver).toHaveBeenCalledWith(1, undefined);
        });

        it('debería retornar respuesta de error cuando el conductor no existe', async () => {
            vi.mocked(repo.deleteDriver).mockResolvedValue(deleteError);

            const result = await useCases.deleteDriver(999);

            expect(result.error).toBeDefined();
            expect(result.message).toBeUndefined();
        });

        it('debería propagar la excepción cuando el repositorio lanza un error de red', async () => {
            vi.mocked(repo.deleteDriver).mockRejectedValue(new Error('Network error'));

            await expect(useCases.deleteDriver(1)).rejects.toThrow('Network error');
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            vi.mocked(repo.deleteDriver).mockResolvedValue(deleteSuccess);

            await useCases.deleteDriver(1, 5);

            expect(repo.deleteDriver).toHaveBeenCalledWith(1, 5);
        });
    });
});
